### DISCUSSION TITLE
Thumbnails sheet

### DISCUSSION THREAD

**Discussion by sergeevabc** [libvips/libvips#3954] [Q&A]
I like the idea of having a [sheet with thumbnails][1] of images with the corresponding filenames. Thus, we save time: instead of looking through all the images in search of the right one, just look at that sheet. But how to generate one with VIPS?

[1]: https://patdavid.net/2013/04/using-imagemagick-to-create-contact/

**Comment by jcupitt** [comment]
Hello @sergeevabc,

You can do this by thumbnailing all the files in a directory, then assembling them with `arrayjoin`.

For example:

```bash
#!/bin/bash

# expand a relative path
dir=$(realpath -s $1)
thumbnails=$dir/.thumbnails

mkdir -p $thumbnails
rm -rf $thumbnails/*
rm $dir/contact-sheet.jpg
echo "generating thumbnails ..."
parallel vipsthumbnail {} -o $thumbnails/%s.jpg --size 256x256 ::: $dir/*
echo "writing $dir/contact-sheet.jpg ..."
vips arrayjoin "$(echo $thumbnails/*)" $dir/contact-sheet.jpg \
        --across 10 --halign centre --valign centre --shim 3
```

And run with:

```
$ contact-sheet waves
```

To scan a directory and make a contact sheet of any recognised images, for example:

![contact-sheet](https://github.com/libvips/libvips/assets/580843/ead7b565-9712-4dff-9333-9240e158f5f6)

You'd probably want to add support for filenames with spaces in, and perhaps subdirectories. You could use `text` to add a caption to each image.

Of course your file manager can probably do this too.

![image](https://github.com/libvips/libvips/assets/580843/6217d334-48ca-4bd0-bddd-24971ab37568)

**Reply by sergeevabc** [reply]
Thank you for your input, @jcupitt.
Alas, I didn't get how to add text labels.

Just in case, I'll leave here a Windows batch file (thumbs.cmd) that uses [ImageMagick][1] 7 and [Jpegli][2].

``` batch
@echo off

montage -verbose -define jpeg:size=512x512 ^
-label "%%f\n%%G, %%B" -font Meslo-LG-S-Regular -pointsize 10 -fill #222222 ^
-auto-orient -colorspace RGB -thumbnail 256x256^> *.jpg -colorspace sRGB -strip ^
-background #dfdfdf -gravity center -extent 256x256^>+0+0 ^
-background #f0f0f0 -geometry 256x256^>+10+10 -tile 3x ^
miff:- | magick - -trim +repage -bordercolor #f0f0f0 -border 20 -depth 8 -quality 0 thumbs.png

cjpegli thumbs.png thumbs.jpg -q 85

del thumbs.png
```

Source: [images.zip](https://github.com/libvips/libvips/files/15242197/images.zip)

Output:

![thumbs](https://github.com/libvips/libvips/assets/373718/041410c7-4fbc-41f0-b9d2-d4a934c5221c)

[1]: https://imagemagick.org/
[2]: https://github.com/libjxl/libjxl

**Comment by jcupitt** [comment]
You'd make a separate script to make a thumbnail with a label, perhaps `make-thumb`:

```bash
#!/bin/bash

path=$(realpath -s $1)
file=$(basename $path)
name="${file%.*}"
dir=$(dirname $path)
thumbnails=$dir/.thumbnails

mkdir -p $thumbnails

width=$(vipsheader -f width $path)
height=$(vipsheader -f width $path)
bytes=$(stat -c %s $path)
txt="$file
${width}x${height}, $bytes"
vips text $thumbnails/$$-2.v "$txt" --font "courier 12" --rgba --align centre

vipsthumbnail $path -o $thumbnails/$$-1.v --size 256x256
vips gravity $thumbnails/$$-1.v $thumbnails/$$-3.v centre 256 256 \
  --background 223
vips gravity $thumbnails/$$-3.v $thumbnails/$$-4.v north 256 300 \
  --background 240

text_width=$(vipsheader -f width $thumbnails/$$-2.v)
text_height=$(vipsheader -f height $thumbnails/$$-2.v)
vips composite2 $thumbnails/$$-4.v $thumbnails/$$-2.v \
  $thumbnails/$name.jpg over \
  --x $(((256 - text_width) / 2)) --y $((300 - text_height - 5))

rm $thumbnails/$$-*.v
```

Then modify the contact sheet thing above to be:

```bash
#!/bin/bash

# expand a relative path
dir=$(realpath -s $1)
thumbnails=$dir/.thumbnails

mkdir -p $thumbnails
rm -rf $thumbnails/*
rm -f $dir/contact-sheet.jpg
echo "generating thumbnails ..."
parallel ./make-thumb {} ::: $dir/*
echo "writing $dir/contact-sheet.jpg ..."
vips arrayjoin "$(echo $thumbnails/*)" $dir/contact-sheet.jpg[Q=90] \
  --across 3 --halign centre --valign centre --shim 20 --background 240
```

To make eg.:

![contact-sheet](https://github.com/libvips/libvips/assets/580843/6ca6566c-d3e9-4e2e-8148-ceb1b67cbaaa)

You'd need to rework it a bit for windows, of course.

**Reply by jcupitt** [reply]
Alternatively, redo this in python. It'd be shorter, simpler, would work on windows and linux, and it wouldn't struggle with spaces in filenames haha.

**Reply by sergeevabc** [reply]
John, if we start requiring a user to install and learn how to use a development environment (e.g. Python) in order to achieve results that are expected to be achieved via the console tool itself (e.g. Vips), we will end up with the fact that it is easier to launch a heavy but intuitive graphics editor. In particular, it is high time to refine `arrayjoin` so that it accepts wildcards, because with a large number of images, they cannot be mentioned in one line.

**Reply by jcupitt** [reply]
The libvips CLI is deliberately very basic -- you can run single operations, and that's it. The idea is that you should use a proper scripting language for anything fancy.

imagemagick (for example) has a complex CLI interface, almost a complete programming language, and I think that's a mistake. They have to maintain it, users have to learn it, and even after all that effort it's still pretty bad. A nice language binding is a better solution.

libvips is designed to be easy to bind, and there are high-quality bindings for most scripting languages. For example, you could install netvips and then call it from powershell, it ought to work well (though I've not tried it myself).

On `arrayjoin`, the CLI interface will start to fail with more than maybe 10,000 images, but I think if you are compositing that many files you should probably be using something better than the CLI!

**Reply by sergeevabc** [reply]
John, on computers running Microsoft Windows XP or later, the maximum length of the string that you can use at the command prompt is 8191 characters ([source][1]). Filenames have long gone beyond 8 characters, for example, the length of `"2024-05-09 Victory day over Nazi Germany.jpg"` string is 44. How many such filenames will fit in a line without taking into account the parameters of VIPS?

8191 / 44 = ~186 images

[1]: https://stackoverflow.com/questions/3205027/maximum-length-of-command-line-string

**Reply by jcupitt** [reply]
Wow, that is tiny. Well, another argument for using python!

**Reply by sergeevabc** [reply]
Python consists of thousands of files, weighs hundreds of megabytes, does not work on Windows 7 without hacks, and requires an effort to be properly configured in a portable way. Now imagine a poor chap from Cuba, Palestine, or Tibet who can't afford expensive graphic editors that require modern hardware, but needs a quick and intuitive tool to make pictures for a relative, event, or lesson. Err, John, what exactly is wrong with the idea of implementing wildcards `*.*` support?

**Reply by jcupitt** [reply]
I think you're maybe overestimating the difficulty of installing python -- it's just a few clicks and a 25mb download. You're right that you do need win8 or newer.

There are a few arguments against globbing in libvips:

1. There's no obvious place to add it. libvips is designed around unix-style globbing in the shell, so there's no simple place to put it. There would need to be a separate binary, perhaps `vipsarrayjoin.exe`, with this feature.
2. It wouldn't always solve the problem. The `arrayjoin` argument needs to list the files in the correct order (row major), so the user needs detailed control over sorting. You can do this with shell globbing with things like eg. `vips arrayjoin "$(echo *.jpeg | sort -t_ -k2g -k1g)"` (you need this to join deepzoom tiles correctly, for example), but how will tasks like that be possible if globbing is in the exe? You'd need to reimplement most of the `sort` command. 
3. Globbing in the command is just bad design. It was a terrible design error to move globbing out of the shell and into programs, it's caused endless trouble. Is it a left-over from CP/M? I'm vague about the history. Let's not go the same way. 

Someone could make a little utility like `contact-sheet.exe`, but I don't think it should be in libvips.

**Comment by jcupitt** [comment]
I had a quick go at a python version, just for fun:

```python
#!/usr/bin/env python

import sys
import os
import glob
import multiprocessing

import pyvips
  
def thumbnail_and_label(filename):
    name = os.path.basename(filename)
    image = pyvips.Image.new_from_file(filename)
    size = os.path.getsize(filename)
    label = pyvips.Image.text(f"{name}\n{image.width}x{image.height}, {size}",
                              font="courier 12",
                              rgba=True,
                              align="centre")
    return pyvips.Image.thumbnail(filename, 256) \
        .gravity("centre", 256, 256, background=223) \
        .gravity("north", 256, 310, background=240) \
        .composite2(label, "over",
                    x=(256 - label.width) / 2,
                    y=300 - label.height - 5) \
        .write_to_buffer(".ppm")
      
for dirname in sys.argv[1:]:
    filenames = sorted(glob.glob(f"{dirname}/*"))

    with multiprocessing.Pool(multiprocessing.cpu_count()) as p:
        thumbs_ppm = p.map(thumbnail_and_label, filenames)
    
    thumbs = [pyvips.Image.new_from_buffer(thumb, "") for thumb in thumbs_ppm]
    sheet = pyvips.Image.arrayjoin(thumbs, 
                                   across=3,
                                   halign="centre",
                                   valign="centre",
                                   shim=20,
                                   background=240)
    sheet \
        .gravity("centre", 
                 sheet.width + 40, 
                 sheet.height + 40, 
                 background=240) \
        .write_to_file(f"{dirname}/contact-sheet.jpg", Q=85)
```

It's about twice the speed of the `montage` command above, on this PC anyway. 

Each thumbnail has to be encoded as ppm to be posted back to the main process, which is unfortunate. A C/C++ implementation could avoid this for a possible speedup.
